src/protocols/secure_channel/DefaultSessionResumptionStorage.cpp

```diff
@@ -52,6 +52,35 @@ CHIP_ERROR DefaultSessionResumptionStorage::Save(const ScopedNodeId & node, Cons
     SessionIndex index;
     ReturnErrorOnFailure(LoadIndex(index));
 
+    for (size_t i = 0; i < index.mSize; ++i)
+    {
+        if (index.mNodes[i] == node)
+        {
+            // Node already exists in the index.  Save in place.
+            CHIP_ERROR err = CHIP_NO_ERROR;
+            ResumptionIdStorage oldResumptionId;
+            Crypto::P256ECDHDerivedSecret oldSharedSecret;
+            CATValues oldPeerCATs;
+            err = LoadState(node, oldResumptionId, oldSharedSecret, oldPeerCATs);
+            // This follows the approach in Delete.  Removal of the old link is
+            // best effort.  Leaks will occur if an entry in the link table does
+            // not have a corresponding entry in the state table.
+            if (err == CHIP_NO_ERROR)
+            {
+                err = DeleteLink(oldResumptionId);
+            }
+            if (err != CHIP_NO_ERROR)
+            {
+                ChipLogError(SecureChannel,
+                             "Unable to delete session resumption link for node " ChipLogFormatX64 ": %" CHIP_ERROR_FORMAT,
```

Comment:
This will log if `LoadState()` returns an error, as well.

Reply:
That is actually the intent.  We need to load from the state table to retrieve the resumption ID in order to remove the entry from the resumption-id-keyed link table.  The implication of being unable to load state is that the resumption-id-keyed entry in the link table won't be removed (this is the storage leak that was happening 100% of the time before this PR).

With the fix, we expect the leak to occur rarely or never, but if it does (or appears to), we want to log it.

Reply:
It would be useful to know which of the two calls failed. Right now we get a similar log for both LoadState failing and DeleteLink failing. If we have 'line numbers enabled' (like linux) then we are ok, but on small devices we may just get an error code and we will not know if line 64 or line 70 failed.

Could we disambiguate that?

Reply:
Ah, right.  I am used to the verbose logging with line numbers being enabled.

Will fix.

Reply:
Done.